launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/LoginServletTest.java

```diff
@@ -0,0 +1,141 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.launchpod;
+
+import static com.google.appengine.api.datastore.FetchOptions.Builder.withLimit;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.security.interfaces.DSAKey;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.gson.JsonParser;
+import com.google.launchpod.data.LoginStatus;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.launchpod.servlets.LoginServlet;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Runs unit tests for the LoginServlet that contains doGet() method.
+ */
+@RunWith(JUnit4.class)
+public class LoginServletTest extends Mockito {
+
+  @InjectMocks
+  private LoginServlet servlet = new LoginServlet();
+
+  @Mock
+  HttpServletRequest request;
+
+  @Mock
+  HttpServletResponse response;
+
+  @Rule // JUnit 4 uses Rules for testing specific messages
+  public ExpectedException thrown = ExpectedException.none();
+
+  private final LocalServiceTestHelper helper = new LocalServiceTestHelper(new LocalUserServiceTestConfig());
+
+  private static final Gson GSON = new Gson();
+  JsonParser parser = new JsonParser();
+
+  private static final String EMAIL = "email";
+
+  private static final String TEST_EMAIL = "123@abc.com";
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  /**
+   * Asserts that doPost() gets the user's email and successfully sends
+   * a corresponding loginStatus object as the response.
+   */
+  @Test
+  public void doGet_GetsCorrectUserLoggedIn() throws IOException {
```

Comment:
this and the below test should each be broken down into two tests since they're each verifying two conditions. you want a separate unit test to test each behavior. one should test that the login status (true/false) is correct and the other is to verify the message.

in the future if you add more fields to the return value, you'll just add new tests - not add verifications to the current ones. if you change any of these behaviors (e.g. change the format of the logout message), you'll only need to edit a single test

Reply:
done.